### PR TITLE
155 feat discover rss providers from url

### LINKED ISSUE #155: feat: discover RSS providers from url

### PR DESCRIPTION
**Pull request by Thechi2000**
Add RSS feed discovery on client side by two methods:
1. Appending predefined pathes to the URL provided by the user.
2. Scraping the HTML of the document at the URL, looking for any URL containing "rss" or "feed", then checking if they are actual feeds.

Closes #155

### CODE REVIEW

**Review by ClementJosso — DISMISSED** [review]
As always very clean code. I have tried theses new settings both with a valid news site (i.e. that has an RSS feed), in which case I got a correct scraping and interesting news, and when I entered an invalid URL I got the error message in red. Great work!

File: lib/models/providers.dart [hunk 1]
```diff
@@ -39,12 +42,80 @@ class _TelegramProviderFactory extends _ProviderFactory {
...
+      ];
+
+      var feeds = (await Future.wait(
+              urls.map((e) => _isRss(e).then((value) => (e, value)))))
+          .where((element) => element.$2);
+
+      if (feeds.isNotEmpty) {
+        return Left(NewsProvider(url: feeds.firstOrNull!.$1.toString()));
+      } else {
+        return const Right(["Unable to find related rss feed"]);
```
Comment (pmdlt) [inline lib/models/providers.dart:105]:
```suggestion
        return const Right(["Unable to find related RSS feed. Please provide the complete URL instead."]);
```

**Review by pmdlt — CHANGES_REQUESTED** [review]
Some minor questions/changes suggested !
Thanks for the job :)

File: lib/models/providers.dart [hunk 2]
```diff
@@ -39,12 +42,80 @@ class _TelegramProviderFactory extends _ProviderFactory {
   }
 }
 
+class _RSSFeedProviderFactory extends _ProviderFactory {
+  const _RSSFeedProviderFactory();
+
+  /// Simple paths which may point to a rss feed.
+  static const paths = [
```
[the remark this answers is not in the document]
Comment (pmdlt) [reply lib/models/providers.dart:49]: But I don't know if it's worth adding to the loop, it affects performance already.

File: lib/models/providers.dart [hunk 3]
```diff
@@ -39,12 +42,80 @@ class _TelegramProviderFactory extends _ProviderFactory {
...
+  static final _rssUrlRegex =
+      RegExp(r'"([^"]*(feed|rss)[^"]*)"', caseSensitive: false);
+
+  /// Checks whether a url is a rss document.
+  ///
+  /// To do so, it requests the url, parses it as a XML file and looks for a <rss></rss> tag.
+  Future<bool> _isRss(Uri url) async {
+    try {
+      var document = XmlDocument.parse((await http.get(url)).body);
+      return document.findAllElements("rss").isNotEmpty;
```
[the remark this answers is not in the document]
Comment (Thechi2000) [reply lib/models/providers.dart:81]: Great idea ! I just added it.

File: lib/models/providers.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (Thechi2000) [reply lib/models/providers.dart:53]: The requests are done in parallel, so it won't affect much.

**Review by ClementJosso — APPROVED** [review]
I see you did dependency injection, great test practice

**Review by pmdlt — APPROVED** [review]
Thanks for the changes and the answers to my questions :)
Looks good to proceed!

**Pull request merged**